File: 09_maximum_slice_problem/max_profit.py
def solution(A):
  n = len(A)
  if n == 0:
    return 0
  
  buy = A[0]
  sell = -1
  max_profit = 0

  for i in A[1:]:
    if i < buy:
      max_profit = max(max_profit, sell - buy)
      buy = i
      sell = i
    else:
      sell = max(i, sell)

  max_profit = max(max_profit, sell - buy)
  if max_profit > 0:
    return max_profit
  else:
    return 0

File: 09_maximum_slice_problem/max_profit.cpp
#include <vector>
#include <algorithm>

using namespace std;

int solution(vector<int> &a) {
  int n = a.size();
  if (n == 0) {
    return 0;
  }
  int ret = 0;
  int least = a[0];
  for (int i = 0; i < n; i++) {
    least = min(least, a[i]);
    ret = max(ret, a[i] - least);
  }
  return ret;
}
